### PR TITLE
[analyzer][clang-tidy][NFC] Clean up eagerly-assume handling

### PR DESCRIPTION
**Pull request by NagyDonat**
This commit is a collection of several very minor code quality improvements. The main goal is removing the misleading "Bin" substring from the names of several methods and variables (like `evalEagerlyAssumedBinOpBifurcation`) that are also applied for the unary logical not operator.

In addition to this, I clarified the doc-comment of the method `evalEagerlyAssumedBinOpBifurcation` and refactored the body of this method to fix the capitalization of variable names and replace an obsolete use of `std::tie` with a structured binding.

Finally, the data member `eagerlyAssumeBinOpBifurcation` of the class `AnalyzerOptions` was completely removed (including a line in clang-tidy that sets it to true), because it was never read by any code.

Note that the eagerly-assume mode of the analyzer is controlled by a different boolean member of `AnalyzerOptions` which is called `ShouldEagerlyAssume` and is defined via the macro magic from `AnalyzerOptions.def`.

### CODE REVIEW

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3749]: These somewhat ugly linebreaks were produced by `git-clang-format`. Perhaps I should try to override it here?

Comment (isuckatcs) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3749]: This declares 2 `static` local variables, right? You can try to split those into 2 different declaration statements and see how `git-clang-format` formats it.

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (isuckatcs) [inline clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3775]:
```suggestion
      const auto &[StateTrue, StateFalse] = State->assume(*SEV);
```

**Review by isuckatcs — COMMENTED** [review]
LGTM apart from some typos.

Maybe we also want to add the unary negation part to the documentation of `ShouldEagerlyAssume` in `AnalyzerOptions.def`.

**Comment by steakhal** [comment]
I want to have a look at this PR tomorrow.

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (NagyDonat) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3773]: The code won't work with `const auto &` because later we assign values to `StateTrue` and `StateFalse` in certain situations.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h
Comment (NagyDonat) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h]: I ended up completely rewriting this comment -- actually this function would work with any boolean expression passed to `Ex` so I don't think that it's helpful to detail what kinds of expression end up being passed to this function.

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3758]: This small clarification is unrelated to everything else -- I just spelled out the type explicitly because that's the preferred style and I'm already cleaning up many things in this function.

In theory I could say `ExplodedNode * const Pred` but I don't think that it's worth to emphasize the `const`ness of the pointer.

Comment (isuckatcs) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3758]: I guess initially it wanted to be `const auto *`. There is no need to make the pointer `const`. Making the pointed value `const` would make sense if it's applicable here.

Comment (NagyDonat) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3758]: > Making the pointed value const would make sense if it's applicable here.

The pointed value cannot be made `const`, we're passing it to functions that modify it. (In general, `ExplodedNode`s are almost never `const` during the analysis, because we can add successor nodes to them in many situations and that modifies one of their data members, the list of successors.)

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (isuckatcs) [inline clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3771]:
```suggestion
      const auto &[TrueTag, FalseTag] = getEagerlyAssumeBifurcationTags();
```
Here the old pattern was also a `const std::pair<..., ...> &`.

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3750]: I'd prefer class static const members for the tags over a function like this.

**Review by steakhal — COMMENTED** [review]
Looks good to me overall.

File: clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def
Comment (NagyDonat) [reply clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:303]: Good point, if you implement it in a follow-up non-NFC commit, I'd be happy to review it. (Alternatively, we could create a todo note / github ticket etc. to record this.)

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (NagyDonat) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3786]: Also a very good point, I think we can skip this node generation if the boolean value is not ambiguous. I don't want to squeeze this change into this NFC commit, because I can easily imagine that skipping a superfluous `ExplodedNode` could cause functional changes in some rare corner case. However, it would be nice and not too difficult to fix this in a follow-up commit.

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (NagyDonat) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3750]: I thought about that, but I don't like global variables (including static members) with not entirely trivial constructors or destructors. I don't think that the initializer of `SimpleProgramPointTag` can cause actual problems (it only does some string concatenation), but it's still a code smell IMO.

Also, I felt that this "get both tags in a single pair, then assign them to short local aliases" interface is better than having two static data members which need to have long names, and referring to e.g. `&EagerlyAssumeBifurcationTrueTag` or something similar.

I'd prefer to keep the current implementation, but I can change it if you strongly prefer a static const member.

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (isuckatcs) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3786]: I don't think that passing the tag here is an issue. The process for creating the node is "eager assumption" and in this case it only creates one branch. 

The tags in the egraph tell the user how the analyzer reached the conclusion to create that node and they are also useful for serving logging/debugging purposes. For example if the egraph is wrong at the node with this tag, you immediately know where the issue is. If there is no tag and the analyzer only creates one branch, it might be difficult to figure out where the node was created.

The only issue might be the wording, which says "Eagerly Assume" even though the analyzer _knows_ the value of the condition, but once again the creation of the node is the result of "eager assumption", so I guess it's fine to leave it like this.

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (isuckatcs) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3750]: After a quick look at the codebase, the pattern seems to be that we create these tags at the point where they are needed as `static` local variables. To not break this pattern, we should either change all of them to be class members, or leave all of them as they are.

Also by being local variables, their visibility is limited to the point where they are needed. I don't think it makes sense to give them class-wide visibility. Also when one reads through the codebase, most of the time they can immediately see what the tag is, while as a class member they would need to first find the definition (i.e.: 1 more click, or pressing 1 more button). The latter reasoning can also be applicable for debugging.

Another thing to keep in mind is that we have some additional logic like the one in `ConditionBRVisitor::VisitNodeImpl()`, which take advantage of these tags, so removing the tags is not an NFC and we might end up rewriting other logic too.

Fun fact: `ProgramPoint` comparison also takes advantage of the tags and IIRC removing the tag in [`a47ec1b`](https://github.com/llvm/llvm-project/commit/a47ec1b79797c8c48c44f9ddf36fb82f8d87229d) might lead to an infinite loop in core engine, when 2 epsilon points follow each other, but it's been a while, so I don't remember it exactly.

Tldr, I would keep the tags as they are for now.

**Review by isuckatcs — APPROVED** [review]
I'm happy with the current state of the patch, so from my end it is approved.

**Comment by NagyDonat** [comment]
@steakhal May I merge this commit?

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
Comment (steakhal) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:3786]: I'd be fine with having a different tag there. Actually, that would be really useful at times - e.g. with tooling around egraphs that is not present upstream. So having a tag is not an issue. Having a tag that lies about that we took an eager decision here is the problem. I agree that this shouldn't be done part of this PR. And I'm also fine leaving this alone in the future. I just wanted to raise this never the less.

File: clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def
Comment (steakhal) [reply clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:303]: I don't get your response here. I thought we want to update the doc comments here to reflect what the function does. I raised that I think implicit conversions also trigger this code path (and do eager splits). Consequently, I figured that this behavior should be also mentioned here and now.

If bool conversions don't trigger this function, then of course my suggestion is void.
I didn't mean to suggest function changes by my comment.

About the TODO notes: I wish we had some lightweight list of things that we as maintainers agree that is easy to do, we have a clear NFC or semantic change in mind but nobody has the time doing it. I have in mind a lot of NFC refactors, API cleanups where I never get to, but I wish would be done.

**Review by steakhal — APPROVED** [review]
LGTM, modulo single inline unresolved comment thread.

File: clang-tools-extra/clang-tidy/ClangTidy.cpp
Comment (whisperity) [inline clang-tools-extra/clang-tidy/ClangTidy.cpp:461]: Is this a functionality loss from Tidy's perspective? AFAIK, there is no way to set CSA analyser engine options when invoked through Tidy.

File: clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def
Comment (whisperity) [inline clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:307]: I know this part of the change is only because of formatting, but here we are saying that _"the downside [of eagerly bifurcating the path] is that it eagerly bifurcates paths"_, which has a very low information value.

File: clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def
Comment (NagyDonat) [reply clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:303]: Oh, I see, I misunderstood your comment. 

Currently implicit bool conversions _do not_ trigger the eagerly-assume code path, and I automatically assumed that you know this and your suggestion says that we should modify this behavior.

In fact I still think that it might be worthwhile to implement this in a follow-up commit (although I don't think that I'll work on this soon); but then it's clear that there is nothing to do with this PR at this point.

> About the TODO notes: I wish we had some lightweight list of things that we as maintainers agree that is easy to do, we have a clear NFC or semantic change in mind but nobody has the time doing it. I have in mind a lot of NFC refactors, API cleanups where I never get to, but I wish would be done.

Nowdays I tend to record these ideas for myself by creating tickets in our internal system; but perhaps I should switch to creating github issues because that would be visible for other contributors as well. Creating a github ticket is not as lightweight as e.g. writing a comment, but I think that it'd be still acceptable. Perhaps we should create a tag/group/whatever for this kind of issue to be able to list them easily.

File: clang-tools-extra/clang-tidy/ClangTidy.cpp
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/ClangTidy.cpp:461]: Clang-Tidy can set CSA options, I recently added support for it for `--verify-config` in #109523, where I had checked which syntax it has, and if the syntax worked beforehand.

Comment (NagyDonat) [reply clang-tools-extra/clang-tidy/ClangTidy.cpp:461]: No, this is not functionality loss from Tidy's perspective, because this data member is dead, its value doesn't influence anything. The `eagerly-assume` feature is controlled by `AnalyzerOptions.ShouldEagerlyAssume` (which is true by default).

File: clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def
Comment (NagyDonat) [reply clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:307]: Fair point, but I think it's still useful to highlight that bifurcating paths _is_ a downside, not a helpful effect.

Comment (whisperity) [reply clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:307]: My problem is that it does not explain well **why** it is a downside, considering we have a strong claim _"increases analysis precision"_ in the other sentence.

File: clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def
Comment (steakhal) [reply clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:303]: Thanks for the explanation.

File: clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def
Comment (isuckatcs) [reply clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:307]: > "the downside [of eagerly bifurcating the path] is that it eagerly bifurcates paths"

I think in this case it's more like _"the downside [of eagerly assuming the value of a bool expression] is that it eagerly bifurcates paths"_

> My problem is that it does not explain well why it is a downside

I think in this case the **why** is that _"it eagerly bifurcates paths"_.

Comment (whisperity) [reply clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:307]: > I think in this case it's more like _"the downside [of eagerly assuming the value of a bool expression] is that it eagerly bifurcates paths"_

I do not think there is a significant difference here, because the help text, as it stands now, immediately tries to explain what a "bifurcation" is (which one may or may not understand that this is actually the definition of that term) by saying _"splits the state to separate […] true and […] false"_.

So we say that turning this option on will cause bifurcation…

> I think in this case the **why** is that _"it eagerly bifurcates paths"_.

…and then go on and say that the bifurcation (which is the only immediate effect of turning this option on) is a downside.

That still does not explain to me why this would be a problem, and what trade-offs I am making when the benefit is the increased precision. Will it cause more CPU use? Will it make the analysis run longer? Or, conversely, will it actually make it run shorter? There is this claim about a behavioural change with no mentions of the actual observeable difference.

And, remember, these documentation tidbits exist for the user, as it is dumped with the `--help` flags and similar. Users will not know the intimacies of the analyser's implementation!

**Pull request merged**
